Adapterize storage for ActionCable

maclover7 opened the pull request:
This is an attempt to adapterize the storage adapters for ActionCable. Currently, the only option is Redis.

Key Changes:
- The default will remain to be Redis for the time being for ease of use via `ActionCable::SubscriptionAdapter::Redis`.
- `ActionCable::SubscriptionAdapter::Base` is introduced to allow other people to write their own adapters.
- `adapter` configuration option is permitted in the yml file, to allow different adapters to be specified.
- To be storage adapter agnostic, the ActionCable yml configuration file is being moved from `config/redis/cable.yml` to just `config/cable.yml`.

maclover7 committed:
Adapterize ActionCable storage and extract behavior

rafaelfranca commented:
Awesome! I think we should go further and investigate a PosgreSQL pubsub adapter. Adapterizing when we only have one adapter is BDUF and we prefer to avoid that.

maclover7 commented:
Ok - do we have support for this through the ActiveRecord DB adapter, or would we need to have to interact manually via the PG gem?

kuldeepaggarwal commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,6 +29,22 @@ def channel_class_names
+        adapter_const = "ActionCable::StorageAdapter::#{adapter.capitalize}"
+
+        if Object.const_defined?(adapter_const)
+          adapter_const.constantize
```
We can use `Object.const_get adapter_const` as we are already checking for `Object.const_defined?(adapter_const)`

maclover7 replied:
Would this let us default to the Redis adapter? That's why I'm checking to see if the provided adapter is defined first, because then we bail out of that and send back the Redis adapter.

pixeltrix commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,6 +29,22 @@ def channel_class_names
+        else
+          ADAPTER_BASE::Redis
+        end
+      end
```
I think it'd be simpler and clearer to just raise if the constant is missing and then you can just implement the method like this:

``` ruby
def storage_adapter
  adapter = config_options.fetch('adapter') { 'redis' }
  "ActionCable::StorageAdapter::#{adapter.camelize}".constantize
end
```

That way it doesn't just magically use Redis if you make a typo in your config file

maclover7 replied:
What would providing the block do for us? Not sure if that's how you're defaulting to Redis.

pixeltrix replied:
The block is called when there is no key 'adapter' set in the options hash

pixeltrix commented:
Is `StorageAdapter` the right name to use? This is mainly for managing pubsub notifications right?

seuros commented:
What about using `PubSubAdapter` ?

maclover7 commented:
IMHO I prefer using StorageAdapter -- the adapter is "storing" the information.

maclover7 committed:
Refactor storage_adapter

maclover7 committed:
Add Postgres adapter

kaspth commented on actioncable/lib/action_cable/storage_adapter/base.rb:
```diff
@@ -0,0 +1,22 @@
+      # Storage connection instance used for pubsub.
+      def pubsub
+        raise NotImplementedError
+      end
```
Think we'd be better off making `pubsub` private and enforce the expected interface with:

```ruby
delegate :publish, :subscribe, :unsubscribe, to: :pubsub
```

maclover7 replied:
IMHO, you shouldn't be calling `publish`, `subscribe`, etc. directly on the adapter -- the adapter is simply the connection holder, not the means of communication with the backend directly.

kaspth commented on actioncable/lib/action_cable/storage_adapter/base.rb:
```diff
@@ -0,0 +1,22 @@
+      # Storage connection instance used for broadcasting. Not intended for direct user use.
+      def broadcast
+        raise NotImplementedError
+      end
```
`broadcast` feels off, as it just returns a dedicated `pubsub` for broadcasting. Would do:

```ruby
def publish_broadcast(*args) # :nodoc:
  broadcaster.publish(*args)
end
```

Note: adding `# :nodoc:` means we don't have to write that it's not intended for direct user use.

kaspth commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -24,11 +24,12 @@ class Railtie < Rails::Engine # :nodoc:
-          self.redis = Rails.application.config_for(redis_cable_path).with_indifferent_access
+        if (config_path = Pathname.new(app.config.paths["config/cable"].first)).exist?
+          # TODO: remove this comment -- anyway, self is an instance of ActionCable::Server::Configuration
+          self.config_opts = Rails.application.config_for(config_path).with_indifferent_access
```
`config_opts` and `@server.config.config_opts` reads off. Can `@server.config.cable` work?

matthewd committed:
Pull the action methods directly onto the adapter

matthewd committed:
Don't execute callbacks on our main listener thread

matthewd committed:
Blanket unsubscribe doesn't appear to be used

matthewd committed:
Listener no longer needs to be a singleton

We now only create one adapter instance for the server, so it can hold
the listener. This in turn allows the listener to get the PG connection
from the adapter, which will be a good place to allow more flexible
configuration.

maclover7 committed:
Tests passing and small refactoring

maclover7 committed:
config_opts => cable, per @kaspth

maclover7 committed:
All Redis deps are now optional, Postgres --> PostgreSQL adapter

maclover7 committed:
ActionCable::StorageAdapter ==> ActionCable::SubscriptionAdapter

maclover7 committed:
Small PostgreSQL adapter refactors / cleanup

- Escape the channel name when subscribing in PG
- Refactor popping the queue to make it easier to read

maclover7 committed:
Add em-hiredis and redis as default Gems for new applications

maclover7 commented:
This looks ready to go from my end - ping @matthewd and @rafaelfranca for review. :)

dhh commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -0,0 +1,104 @@
+        end
+      end
+
+      private
```
Nix CR after `private` and indent everything below.

dhh commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -0,0 +1,104 @@
+      class Listener
+        def initialize(adapter)
+          @adapter = adapter
+          @subscribers = Hash.new {|h,k| h[k] = [] }
```
`Hash.new { |h,k| h[k] = [] }`

dhh commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,46 @@
+      end
+
+      private
+
```
Nix CR, indent private methods.

dhh commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,46 @@
+      private
+
+      # The EventMachine Redis instance used by the pubsub adapter.
+      def hi_redis_conn
```
I'd go with `subscription_redis_connection`. We generally avoid abbreviations like `conn` (or we certainly should) and "hi/em" is a implementational detail.

dhh replied:
Actually, I like `redis_connection_for_subscriptions` even better.

dhh commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,46 @@
+        end
+      end
+
+      def redis_conn
```
`redis_connection_for_broadcasts`

dhh commented on actioncable/lib/action_cable/connection/base.rb:
```diff
@@ -49,7 +49,7 @@ class Base
 
       attr_reader :server, :env, :subscriptions, :logger
-      delegate :worker_pool, :pubsub, to: :server
+      delegate :worker_pool, :adapter, to: :server
```
I agree with the earlier comments that calling this `adapter` rather than `pubsub` is a step backwards. Adapter is a generic pattern. We could have adapters for all sorts of things. This is specifically an adapter that provides the pubsub backend we need. So let's stick with the `pubsub` name. Our use of `broadcast` vs `publish` is more just to stay within the cable metaphor. It's still the same thing.

dhh commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,16 @@ def channel_class_names
+
+      # Returns constant of subscription adapter specified in config/cable.yml
+      # If the adapter cannot be found, this will default to the Redis adapter
+      def subscription_adapter
```
This would then be `pubsub_adapter` as well.

dhh commented:
I don't see any tests that exercise the new PGSql adaper. How has this been tested so far?

dhh commented:
BTW, this is really nice. Thanks for working on this @maclover7! It would be great to provide some basic benchmarks for people to contrast the different adapters.

kaspth commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,16 @@ def channel_class_names
+      # Returns constant of subscription adapter specified in config/cable.yml
+      # If the adapter cannot be found, this will default to the Redis adapter
+      def subscription_adapter
+        # Defaults to redis if no adapter is set
```
Superflous, already explained just above.

kaspth commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,16 @@ def channel_class_names
+      def subscription_adapter
+        # Defaults to redis if no adapter is set
+        adapter = cable.fetch('adapter') { 'redis' }
+        adapter = adapter.camelize
```
Spare the line; move `camelize` to just above.

kaspth commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,16 @@ def channel_class_names
+        # Defaults to redis if no adapter is set
+        adapter = cable.fetch('adapter') { 'redis' }
+        adapter = adapter.camelize
+        adapter = "PostgreSQL" if adapter == "Postgresql"
```
We could just call the adapter `Postgres` and spare this munging (otherwise to do this right, we'd need a more advanced registry, which is too much).

maclover7 replied:
I think we went with "PostgreSQL" to have it match AR conventions. AR does the same thing, [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/connection_specification.rb#L37)

kaspth replied:
Thanks for taking me to school, all good :+1:

> Den 16. jan. 2016 kl. 16.39 skrev Jon Moss <notifications@github.com>:
> 
> In actioncable/lib/action_cable/server/configuration.rb <https://github.com/rails/rails/pull/22950#discussion_r49933190>:
> 
> > @@ -29,7 +29,16 @@ def channel_class_names
> >            Pathname.new(channel_path).basename.to_s.split('.').first.camelize
> >          end
> >        end
> > +
> > +      # Returns constant of subscription adapter specified in config/cable.yml
> > +      # If the adapter cannot be found, this will default to the Redis adapter
> > +      def subscription_adapter
> > +        # Defaults to redis if no adapter is set
> > +        adapter = cable.fetch('adapter') { 'redis' }
> > +        adapter = adapter.camelize
> > +        adapter = "PostgreSQL" if adapter == "Postgresql"
> I think we went with "PostgreSQL" to have it match AR conventions. AR does the same thing, here <https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/connection_specification.rb#L37>
> —
> Reply to this email directly or view it on GitHub <https://github.com/rails/rails/pull/22950/files#r49933190>.
> 

--
Kasper

kaspth commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,16 @@ def channel_class_names
+        adapter = cable.fetch('adapter') { 'redis' }
+        adapter = adapter.camelize
+        adapter = "PostgreSQL" if adapter == "Postgresql"
+        return "ActionCable::SubscriptionAdapter::#{adapter}".constantize
```
Remove superflous `return`.

maclover7 committed:
Fix code review comments

- adapter -> pubsub (re)rename internally
- Change variable names to match method names
- Add EventMachine `~> 1.0` as a runtime dependency of ActionCable
- Refactor dependency loading for adapters

maclover7 commented:
Thanks for the review, @dhh!

- Tried to get benchmarks running using https://github.com/observing/thor and https://github.com/M6Web/websocket-bench, but couldn't get everything wired up properly. Any advice here?
- Have been testing the PG adapter using the "ActionCable Demo" video application. You can find it [here](https://github.com/maclover7/ac-test), though I think I have some local commits on it that I haven't pushed up.
- Not completely sure how AC's test suite is setup to test out things, since there are a lot of stubs, would you be able to offer some advice about how to test out the adapter properly?

kaspth commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -0,0 +1,103 @@
+begin
+  require 'pg'
+rescue Gem::LoadError => e
+  raise Gem::LoadError, "You are trying to use the PostgreSQL ActionCable adapter, but do not have the proper gems installed. Add `gem 'pg'` to your Gemfile (and ensure its version is at the minimum required by ActionCable)."
```
When we aren't referring to the library's constant, there's a space in the name: Action Cable, Active Record etc.

kaspth commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,45 @@
+  require 'redis'
+rescue Gem::LoadError => e
+  raise Gem::LoadError, "You are trying to use the Redis Action Cable adapter, but do not have the proper gems installed. Add `gem 'em-hiredis'` and `gem 'redis'` to your Gemfile (and ensure its version is at the minimum required by Action Cable)."
+end
```
Would consider moving this into a class macro in the adapters themselves, so we can share the raise and message behavior.

How about:

```ruby
module ActionCable
  module SubscriptionAdapter
    class Base
      def self.require_gems!(*gems)
        gems.each { |dependency| require dependency }
      rescue Gem::LoadError => error
        raise error, "You're trying to use the #{name} Action Cable pubsub adapter, but you're missing " \
          "#{gems.join(', ')} from your Gemfile."
      end
    end
    
    class Postgres < Base
      require_gems! 'pg'
    end
    
    class Redis < Base
      require_gems! 'em-hiredis', 'redis'
    end
  end
end
```

maclover7 replied:
:+1: I like this abstraction.

kaspth commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -0,0 +1,103 @@
+              end
+
+              pg_conn.wait_for_notify(1) do |chan, pid, message|
+                @subscribers[chan].each do |callback|
```
Chan can subscribe and @chancancode! :grin:

kaspth commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -0,0 +1,103 @@
+          end
+        end
+
+        def unsubscribe_to(channel, callback)
```
Would rename to `unsubscribe_from`

kaspth commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,45 @@
+
+module ActionCable
+  module SubscriptionAdapter
+    class Redis < Base
```
Wouldn't it be better to just mark these as `# :nodoc:` don't see anything here that users should be calling directly.

kaspth commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,45 @@
+        broadcast_redis_connection.publish(channel, payload)
+      end
+
+      def subscribe(channel, callback, success_callback = nil)
```
Since there's a variable called `success_callback` that leaves `callback`'s intent up in the air. Starting to question the suffix entirely too. Perhaps the more succinct `subscriber` and `notifier`, or `on_subscribe` and `on_success` could capture what you're trying to convey.

kaspth commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,45 @@
+      end
+
+      private
+        # The EventMachine Redis instance used by the pubsub adapter.
```
Doesn't add much; :scissors:.

kaspth commented on railties/lib/rails/generators/app_base.rb:
```diff
@@ -339,6 +340,15 @@ def psych_gemfile_entry
 
+      def cable_gemfile_entry
+        return [] if options[:skip_action_cable]
+        comment = 'ActionCable dependency for the Redis adapter'
```
`Action Cable`

kaspth commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -24,11 +24,12 @@ class Railtie < Rails::Engine # :nodoc:
-        if (redis_cable_path = Pathname.new(app.config.paths["config/redis/cable"].first)).exist?
-          self.redis = Rails.application.config_for(redis_cable_path).with_indifferent_access
+        if (config_path = Pathname.new(app.config.paths["config/cable"].first)).exist?
+          # self is an instance of ActionCable::Server::Configuration
```
Doesn't add much, :scissors:.

dhh commented:
This collapses the subscription and broadcasting redis instances. I'm pretty sure they were split for a reason. Asking @lifo what the story is.

dhh commented on actioncable/lib/action_cable/server/broadcasting.rb:
```diff
@@ -46,7 +39,7 @@ def initialize(server, broadcasting)
           def broadcast(message)
             server.logger.info "[ActionCable] Broadcasting to #{broadcasting}: #{message}"
-            server.broadcasting_redis.publish broadcasting, ActiveSupport::JSON.encode(message)
+            server.pubsub.broadcast broadcasting, ActiveSupport::JSON.encode(message)
```
Actually, this is the only place that mistake is made. We need to use the broadcasting redis here. Not the evented pubsub redis.

maclover7 replied:
`pubsub` is just the instance of `ActionCable::SubscriptionAdapter::**`. https://github.com/rails/rails/pull/22950/files#diff-32900b7f3328eeea82a1443ed88197b0R50

dhh replied:
Ah, yes, my bad. I should have read it all the way down. All good.

maclover7 commented:
How does this merge the two instances into one? `broadcast_redis_connection` and `subscription_redis_connection` are separate instances -- one is provided by `EM::Hiredis` and the other through `Redis` itself.

maclover7 commented:
Needed to rebase.

dhh commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -5,9 +5,9 @@ module Server
+      attr_accessor :channels_path
       attr_accessor :disable_request_forgery_protection, :allowed_request_origins
-      attr_accessor :url
+      attr_accessor :adapter, :cable, :url
```
Should go with :pubsub for the instance of the adapter here too, imo. The adapter is the abstract pattern, the pubsub is the instance.

maclover7 replied:
`adapter` is just returning whatever the user inputs in their `config/cable.yml` file for the adapter name, so it'll just be a string that says `redis`, `postgresql`, or something else.

dhh replied:
Ah, yes. This is configuration. I should pay more attention.

rafaelfranca replied:
Where it is being used? I didn't saw any call of this `adapter` method.

kaspth replied:
It's from `config/cable.yml` where adapter is assigned.

rafaelfranca replied:
But where is it being assigned? As far I can see the YML configuration is assigned to `cable` but there is not assignment to adapter.

kaspth replied:
Ah, right. I'm actually not sure.

matthewd replied:
Agree.. I think `:adapter` here is vestigial. The "adapter" we care about lives inside `cable`.

dhh commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,37 @@
+            redis.on(:reconnect_failed) do
+              @logger.info "[ActionCable] Redis reconnect failed."
+              # logger.info "[ActionCable] Redis reconnected. Closing all the open connections."
+              # @connections.map &:close
```
Was this commented out before as well?

maclover7 replied:
yes

dhh commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,37 @@
+
+      private
+        def subscription_redis_connection
+          @hi_redis_conn ||= EM::Hiredis.connect(@server.config.cable[:url]).tap do |redis|
```
Let's change the variable names to match the method names.

dhh commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,37 @@
+        end
+
+        def broadcast_redis_connection
+          @redis_conn ||= ::Redis.new(@server.config.cable)
```
Var name => method name.

dhh commented:
Okay, looking here. Would be great to get more people to try out the pgsql adapter, but at least now it's available and we can refine it.

maclover7 commented:
Cool :+1: -- sounds like we're ready to merge then?

rafaelfranca commented:
Squash the commits please so we can merge

rafaelfranca commented on actioncable/actioncable.gemspec:
```diff
@@ -21,11 +21,13 @@ Gem::Specification.new do |s|
   s.add_dependency 'actionpack', version
 
   s.add_dependency 'coffee-rails',     '~> 4.1.0'
+  s.add_dependency 'eventmachine'
```
`~>`

rafaelfranca commented on actioncable/lib/action_cable/subscription_adapter/base.rb:
```diff
@@ -0,0 +1,33 @@
+
+      def self.require_gems!(*gems)
+        begin
+          gems.each { |gem| require gem }
```
WE should use the same version of pg that we use in AR. And we should call `gem 'pg'` like we do in AR.

rafaelfranca commented on actioncable/lib/action_cable/engine.rb:
```diff
@@ -24,11 +24,11 @@ class Railtie < Rails::Engine # :nodoc:
       ActiveSupport.on_load(:action_cable) do
-        if (redis_cable_path = Pathname.new(app.config.paths["config/redis/cable"].first)).exist?
-          self.redis = Rails.application.config_for(redis_cable_path).with_indifferent_access
+        if (config_path = Pathname.new(app.config.paths["config/cable"].first)).exist?
```
To this work the file need to be registered. https://github.com/rails/rails/blob/48f5cecdb03f697a710169ccef1885fcaca70788/actioncable/lib/action_cable/engine.rb#L27

maclover7 replied:
Yes - the file is registered on line 27, and Rails generates the file in new applications.

rafaelfranca commented on actioncable/lib/action_cable/subscription_adapter/redis.rb:
```diff
@@ -0,0 +1,37 @@
+          @subscription_redis_connection ||= EM::Hiredis.connect(@server.config.cable[:url]).tap do |redis|
+            redis.on(:reconnect_failed) do
+              @logger.info "[ActionCable] Redis reconnect failed."
+              # logger.info "[ActionCable] Redis reconnected. Closing all the open connections."
```
why this code commented?

maclover7 replied:
was commented originally when I found it

rafaelfranca replied:
:fire: the comments

matthewd commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,25 @@ def channel_class_names
+          raise LoadError, "Could not load '#{path_to_adapter}'. Make sure that the adapter in config/cable.yml is valid. If you use an adapter other than 'postgresql' or 'redis' add the necessary adapter gem to the Gemfile.", e.backtrace
+        end
+        ###
+        adapter.camelize
```
`adapter = adapter.camelize`

matthewd replied:
Fixed in merge

Sija commented on actioncable/lib/action_cable/subscription_adapter/postgresql.rb:
```diff
@@ -0,0 +1,99 @@
+gem 'pg', '~> 0.18'
```
Is this is right place to put `Bundler` definitions?

kaspth replied:
It's not a bundler definition it's aactivating a specific version of a gem. Most commonly I've seen this done with Minitest to ensure any require uses the gem and not the version shipping with Ruby.

Tsumanu replied:
And what if someone does not have this version installed?

kaspth replied:
To use this subscription adapter they should.

koenpunt commented on actioncable/lib/action_cable/server/configuration.rb:
```diff
@@ -29,7 +29,25 @@ def channel_class_names
+      # If the adapter cannot be found, this will default to the Redis adapter.
+      # Also makes sure proper dependencies are required.
+      def pubsub_adapter
+        adapter = (cable.fetch('adapter') { 'redis' })
```
I've seen this syntax more in Rails 5, what's the advantage of `cable.fetch('adapter') { 'redis' }` over `cable.fetch('adapter', 'redis')` ?

kaspth replied:
The advantages are the same as any other times work is deferred. In this case a string allocation is saved. Doesn't matter a huge deal here, since this isn't a hotspot, but still 😁

koenpunt replied:
I understand there's a performance win when there's some actual work done in the block, but when that's not the case, doesn't take the block any memory?

thedarkone replied:
Whats wrong with simple `adapter = cable['adpater'] || 'redis'`? Or is `cable['adapter'] = false` a valid config?

PS: I bet it even performs better than both of the `fetch` versions.

koenpunt replied:
I believe that this point of the code is only accessed once; while initializing, so optimizing is not necessary here. But I'm just curious about `fetch` vs block in general
